### ISSUE TITLE
ref locals can't be re-assigned by reference!

### ISSUE THREAD

**Issue by MohammadHamdyGhanem** [dotnet/csharplang#1202]
I have a function like this:
```
ref string FindStudent(string Name) 
{
}
```
so, I use it like this:
ref var std = ref FindStudent("Adam");
now, I try to reuse the std variable in another call to the function, but c# refuses all of the next statements:
`std = ref FindStudent("John");`
`ref std = ref FindStudent("John ");`
`ref std = FindStudent("John");`
The only accepted way is:
`std = FindStudent("John");`
but it returns a value not a ref return!
Please fix this. Also I suggest to get red of the second ref before the call of the function. It is sufficient to write:
`ref var std = FindStudent("Adam");`
Also, why can a ref variable take a direct value?.. This is illogical by definition!. so, the statement:
`std = FindStudent("John");`
should return a ref .
Thanks.

**Comment by DavidArno** [comment]
According to [this proposal](https://github.com/dotnet/csharplang/blob/master/proposals/ref-local-reassignment.md), that feature was added to v7.2. But according to #996, it was moved out of the 7.2 folder, so presumably didn't get released in time and the proposal needs updating.

So "coming soon" seems to be the status of this request.

**Comment by mikedn** [comment]
> Please fix this

Why? You keep asking for more `ref` related features but the only reasoning you provide is "this is illogical by definition". Illogical according to what kind of logic?

**Comment by MohammadHamdyGhanem** [comment]
@mikedn 

> but the only reasoning 

please read my post again. I talk about two things and you mixed them.

> Illogical according to what kind of logic?

`std = FindStudent("John");`
1. there is a ref local on the hand side
2. there is a ref return on the right side
3. the result is value return!!!!
Conclusion: Illogical!
It should return the ref or C# consider it a syntax error, otherwise forgetting to write the ref keyword will produce different results.
The Question is: 
Why does a high-level compiler need a special operator (=ref) to understand the obvious?
Will the operator (=ref) be able to assign the ref of any data type, without being declared ref? If not, why we need it?

**Comment by orthoxerox** [comment]
`ref` has sufficiently dangerous semantics to warrant explicit mention every time you need it.

**Comment by HaloFour** [comment]
@MohammadHamdyGhanem 

>  Why does a high-level compiler need a special operator (=ref) to understand the obvious?

It doesn't.  The operator is forced on the developer so that they have to think about the fact that they're using `ref` and not normal copy-by-value variables.

**Comment by mikedn** [comment]
> Conclusion: Illogical!

You're jumping to conclusions without trying to understand the facts. The fact is that every time a reference is used without specifying `ref` you are accessing the value stored in the referenced location. That's how references work in C# from day 1. That's also how references work in C++.

> otherwise forgetting to write the ref keyword will produce different results

In programming there are many things that if you forget doing you end up with different results. For all I know you could also forget writing the entire assignment. It's best for a language to minimize the risk of making mistakes but at the same time it needs to stay out of the way and not block valid scenarios. There's nothing wrong or illogical about `std = FindStudent("John");`. It copies a value from one location to another. That's what assignment always did in C#.

> Why does a high-level compiler need a special operator (=ref) to understand the obvious?

You don't seem to understand how things work. A "high-level" compiler doesn't need anything. A compiler implements a language specification. And that language specification is designed to meet certain requirements. In this case the requirement is to make it obvious that references and not values are assigned. @orthoxerox already explain why this is needed.

**Comment by MohammadHamdyGhanem** [comment]
@mikedn 

> The fact is that every time a reference is used without specifying ref you are accessing the value stored in the referenced location. That's how references work in C# from day 1

I think you are making false "facts" , so you can't understand what I am talking about!
When you use "=" between two objects (like arrays) this means a ref assignment whithout any ref keyword! That's how references work in C# from day 1.
Now if the =ref introduced there will be a lot of confusion between = , and =ref.
Also, I hope you can understand this simple true fact:
C# not the true .net compiler, so it can make things easy and clear and logical without affecting the performance of the JIT compiler, because eveything is compiled eventually into standered IL code. Treating C# syntax as if it's a low level c++ syntax is a mistak. 
As I said above, the = operator has mixed roles already and you can't cahange that now, so go on and build on it.

**Comment by HaloFour** [comment]
@MohammadHamdyGhanem 

Wrong, the following has always meant a copy, not a `ref`.  Changing `y` would not affect `x`.

```cs
public void Foo(ref int x)
{
    int y = x;
}
```

The same is true with reference-types:

```cs
public void Foo(ref int[] x)
{
    int[] y = x;
}
```

In this case you're **copying** the reference.

With `ref locals`, now the following is possible:

```cs
public void Foo(ref int[] x)
{
    ref int[] y = ref x;
    y = new int[] { 1, 2, 3 };
}
```

In this case `y` actually points to `x`, and changing `y` also changes `x`, which also changes the caller of `Foo`.  That is new, and that **requires** new syntax.

**Comment by mikedn** [comment]
> When you use "=" between two objects (like arrays) this means a ref assignment whithout any ref keyword! 

No it does not. For the purpose of this `ref` specific discussion an assignment copies a 
"value" from one location to another. The fact the for reference types the "value" is actually an object reference is irrelevant. It's unfortunate that the reference term is overloaded like this but `ref` references and object reference are distinct concepts.

> Also, I hope you can understand this simple true fact:
C# not the true .net compiler, so it can make things easy and clear and logical without affecting the performance of the JIT compiler, because eveything is compiled eventually into standered IL code. 

No, I do not understand "this simple true fact". While `ref` tends to be associated with performance optimizations the fact is that it has particular semantics that cannot be substituted by JIT optimizations. So I really do not know what you're trying to say.

> As I said above, the = operator has mixed roles already and you can't cahange that now, so go on and build on it.

And that's exactly what they did - `ref` is required in many places exactly because of this.

**Comment by MohammadHamdyGhanem** [comment]
@HaloFour 
Thanks for your good answer. I thought the ref notation is made only for ref returns (all samoles on docs abiut ref returns). Your sample showed that it can be used freely with other types. This is confusing as it allows somethings were not possible like this sample:
```
int x = 5;
ref int y = ref x;
y = 8;
Consile.WriteLine(X); // 8
```
this answeres my question:

> Will the operator (=ref) be able to assign the ref of any data type, without being declared ref?

Obviously it does, so now I'm convinced it is a new operator.
Thanks again.

**Comment by MohammadHamdyGhanem** [comment]
now here is a new question:
How can we check for ref equality?
`if (y == x)`
 grantees value equality, so can we do it like this:
`if (y == ref x)`
or like this:
`if (y is ref x)`
or just:
`if (y is x)`
C# doen't support any of these yet!

**Comment by HaloFour** [comment]
@MohammadHamdyGhanem 

> How can we check for ref equality?

Why would you want to?

> C# doen't support any of these yet!

Nor should it unless it can be demonstrated that there is actually a use case for doing so.

**Comment by mikedn** [comment]
> Nor should it unless it can be demonstrated that there is actually a use case for doing so.

Indeed. Especially that such exotic features can be implemented in a library - `bool AreSame<T>(ref T left, ref T right) from System.Runtime.CompilerServices.Unsafe`.

**Comment by MohammadHamdyGhanem** [comment]
Whey you test for ref equality of objects? Sure the need emerges! Programming is full of all possible scenarios.
You add a new feature to the language. Let it be complete. 
I said I'm not a fan of ref returns. But certainly I'm a fan of ref locals!.. They are exactly high level safe pointers!
For example: now I can write a linked list class in a new fashion. Testing for ref equality can prevent circular links or whatever else!
Let's forget about ref returns. Ref locals are a brave new world and we should explore it!
Sorry, but it seems that you opened a door that shouldn't be opened. Sure this will raise up the performance in some areas, so you should provide a full implementation, and let us be creative!

**Comment by MohammadHamdyGhanem** [comment]
Two other questions:
1- Can COM pointers/references converted to a ref local?
2- What if I need an array of refs, or a List:
List<ref int> L;

**Comment by mikedn** [comment]
> Can COM pointers/references converted to a ref local?

Huh?

> What if I need an array of refs, or a List:

It's not possible.

**Comment by HaloFour** [comment]
@MohammadHamdyGhanem 

> Sure the need emerges!

And when it does then you can demonstrate the use case and make a proposal.

> now I can write a linked list class in a new fashion.

I don't believe that's possible, at least not today.  It's not like you can assign `ref` locals to fields.

> Sorry, but it seems that you opened a door that shouldn't be opened.

Maybe you should actually understand the feature and its purpose rather than postulating randomly as to how it could be [ab]used in scenarios for which it was clearly not designed nor intended.

**Comment by MohammadHamdyGhanem** [comment]
`it was clearly not designed nor intended`
Are you sure about  that "clearly" thing?
This is the most mysterious c# feature ever been, and there is no decent sample to show why to use it or who should use it .
Let's wait for the =ref operator and see what can be done with ref locals.
Thanks.

**Comment by qrli** [comment]
@MohammadHamdyGhanem You are confusing the `ref` with pointers. C# has pointers from v1. What you expressed are features and behaviors of pointers. It is like that and remains like that.

However, `ref` is never intended to be a another type of pointer, and we don't need two ways to do the same thing.  It is mainly for performance crazy library developers (who count CPU cycles daily), to expose easier APIs without pointers to normal developers, yet still without `struct` copy cost. It is not for normal developers to do daily job.

**Comment by MkazemAkhgary** [comment]
>This is the most mysterious c# feature ever been, and there is no decent sample to show why to use it or who should use it .

This is common problem for all newbies (including me back then) and it seems mysterious until you understand how things work. 

You need to understand that we have two kinds of data types and two ways to pass variables. In conclusion there are four combinations

1- pass value type by value
2- pass value type by reference
3- pass reference type by value
4- pass reference type by reference

Once you understand difference between these 4 cases you you will get a better picture of whats going on and may realize your questions and assumptions are wrong in first place.

When you assign variable without using `ref` that serves as 2 or 4 depending on data type kind. If you dont pass using `ref` that serves as 1 and 3 depending on data type kind.

**Comment by HaloFour** [comment]
@MkazemAkhgary 

> If you are new programmer try to avoid it. You really really dont have to pass reference type by ref. Let me rephrase. You should not pass variables by reference. Its a neat feature for experienced programmers (no pun intended) and even they have to be careful about using that.

Passing variables by `ref`/`out` isn't a "neat feature for experienced programmers".  It's a normal feature for when you have the use case for doing it.

The rule for "less experienced" developers is exceedingly simple:

* Do you need the function to change the value of the parameter and have it reflected by the caller?  
    * **Yes**: use `ref`/`out`
    * **No**: don't use `ref`/`out`. 

But these are, by far, not esoteric features intended for a minority audience.  Simple examples include `int.TryParse` or `Dictionary.TryGetValue` which I suspect most developers use once in a while.

Now when it comes to `in` parameters, `ref` returns or `ref` locals, then you might have a bit of a point.  Those are largely performance-specific features intended for a relatively narrow audience with all kinds of fun caveats.  Few developers will probably be directly exposed to those features (maybe as consumers of `in` parameters), but many developers will benefit from the performance benefits they enable in the underlying framework.
